lib/maths/CBayesianOptimisation.cc

```diff
@@ -372,6 +398,268 @@ double CBayesianOptimisation::kernel(const TVector& a, const TVector& x, const T
+                if (core::CStringUtils::stringToType(traverser.value(), functionValue) == false) {
+                    LOG_ERROR(<< "Error restoring function value: " << traverser.value());
+                    return false;
+                }
+            } else {
+                LOG_ERROR(<< "Error restoring function mean values. Unexpected tag: "
+                          << traverser.name());
+                return false;
+            }
+        } while (traverser.next());
+        return true;
+    };
+    return restoreParameterValuePair;
+}
+
+bool CBayesianOptimisation::restoreFunctionMeanValues(CBayesianOptimisation::TVectorDoublePrVec& functionMeanValues,
+                                                      core::CStateRestoreTraverser& traverser) {
+    if (traverser.hasSubLevel()) {
+        auto restoreParameterValuePairs = [&functionMeanValues,
+                                           this](core::CStateRestoreTraverser& traverser) {
+            TVectorDoublePrVec newVector;
+            do {
+                if (traverser.name() == FUNCTION_PARAMETER_VALUE_PAIR_TAG) {
+                    if (traverser.hasSubLevel()) {
+                        TVector parameters;
+                        double functionValue;
+                        if (traverser.traverseSubLevel(this->restoreParameterValuePair(
+                                parameters, functionValue)) == false) {
+                            return false;
+                        }
+                        newVector.emplace_back(parameters, functionValue);
+                    } else {
+                        LOG_ERROR("Error restoring function parameter value paris: "
+                                  << traverser.value() << ". Expected to have a sublevel");
+                        return false;
+                    }
+                } else {
+                    LOG_ERROR(<< "Unexpected name for function parameter value pair: "
+                              << traverser.name() << ". Expected name: "
+                              << FUNCTION_PARAMETER_VALUE_PAIR_TAG);
+                    return false;
+                }
+
+            } while (traverser.next());
+            functionMeanValues = std::move(newVector);
+        };
+        traverser.traverseSubLevel(restoreParameterValuePairs);
+
+    } else {
+        LOG_ERROR("Error restoring function mean values: "
+                  << traverser.value() << ". Expected to have a sublevel");
+        return false;
+    }
+    return true;
+}
+
+bool CBayesianOptimisation::acceptRestoreTraverser(core::CStateRestoreTraverser& traverser) {
+    try {
+
+        // Call name() to prime the traverser if it hasn't started
+        traverser.name();
+        if (traverser.isEof()) {
+            //            m_RestoredStateDetail.s_RestoredStateStatus = E_NoDetectorsRecovered;
+            LOG_ERROR(<< "Expected persisted state but no state exists");
+            return false;
+        }
```

Comment:
I'm not sure this is needed. Basically, the first restore you do inside the loop will catch this error and fail.

Reply:
I copied it from `CAnomalyJob.cc`. It isn't needed here?

Reply:
My thinking was that this method shouldn't have this sort of top-level functionality since one might expect it to be called as part of restoring state comprising multiple objects.

As with `CAnomalyJob` my inclination would be a `restore` function on `CBoostedTree`, or as we discussed offline, moving this to the factory object.

Reply:
I agree. I removed the the lines from the code.